### ISSUE TITLE
collection: python_files=* / explicit collection

### ISSUE THREAD

**Issue by blueyed** [pytest-dev/pytest#4476]
I think it would be nice if `pytest =(echo "def test_foo: pass")` would work, i.e. it should consider any explicitly passed in file arguments for collection.  (Zsh creates a temporary file here).

Also using `-o 'python_files=*'` does not appear to help here.

Related/via https://github.com/pytest-dev/pytest/issues/4452.

**Comment by RonnyPfannschmidt** [comment]
im -1 on that one - if its not having a .py extension i don't want to see it collected as python module in core

**Comment by blueyed** [comment]
Fair point.

What about a warning then if `python_files` does not match `*.py` / ends with `.py`?

**Comment by RonnyPfannschmidt** [comment]
that sounds usefull

**Comment by RonnyPfannschmidt** [comment]
sidenote - https://unix.stackexchange.com/questions/482459/specifying-the-file-extension-produced-by-zsh-process-substitution should set the temporary file thing right at least in part - it may need extra variables to make it pytohn-safe

**Comment by anarcat** [comment]
> sidenote - https://unix.stackexchange.com/questions/482459/specifying-the-file-extension-produced-by-zsh-process-substitution should set the temporary file thing right at least in part - it may need extra variables to make it pytohn-safe

this doesn't help for my use case which is a small script that I don't want to have to ship as an entire package (see also #3520)

**Comment by anarcat** [comment]
why was this closed? is this abandoned, or fixed? thanks!

**Comment by blueyed** [comment]
@anarcat a new issue / PR should be created based on https://github.com/pytest-dev/pytest/issues/4476#issuecomment-442598532 probably.  Please do so if you want to "own" / work on it.
This issue by itself appears to be a no-go for pytest itself.

**Comment by sarnold** [comment]
Seriously?  I'm with @anarcat on this one, mainly to satisfy packaging QA tools that yell at you for shipping a "binary" with a .py extension.  If i do that ^^ pytest not only doesn't find the file but refuses to even touch it.  Is the answer really "tell the package QA tool to blow" and always keep the .py extension if you want pytest to check it?  In this case it's not a "trivial"/small script but a an actual daemon script.

**Comment by nicoddemus** [comment]
> im -1 on that one - if its not having a .py extension i don't want to see it collected as python module in core

IIUC the proposal here is for `python_files=*` to be honored by the `python` plugin and try to collect anything, even if they don't have the `.py` extension. I believe the chance for misuse here is minimal. Here's the relevant code:

https://github.com/pytest-dev/pytest/blob/8be9684ab23220cc85abb06e06414961667340b3/src/_pytest/python.py#L173-L182

All it would need is to remove the initial check for `".py"` extension; of course I know the issue here is not one of implementation complexity, but opportunities for misuse; I'm just curious to know what are those opportunities, given that in this proposal the user must explicitly configure pytest to ignore the file extension.

**Comment by blueyed** [comment]
I agree with @nicoddemus here, re-opening for now.
Maybe @RonnyPfannschmidt and/or @asottile can explain/elaborate why this should not be allowed (through explicit configuration).

(@nicoddemus 
The initial issue was for `pytest foo.txt` to work already, without changing `python_files` - i.e. any given file would be collected already, but I can see that the "python" plugin should only handle "python_files" really.)

**Comment by RonnyPfannschmidt** [comment]
at first glance we would have to reconfigure both the python import system and the assertion rewriter to do it correctly, if that's the complexity needed, then i don't want to see it in core a plugin will have to contain it

**Comment by nicoddemus** [comment]
Agreed, if it is more complex than a first glance shows, then we should avoid adding it to the core. 

But if a simple implementation can be shown to support it, I would be OK with adding it to the core.

**Comment by blueyed** [comment]
The assert rewriting uses `python_files` (without validation), so it's not more complex than removing the check that @nicoddemus pointed out.
(As for having a separate config for the assert rewriting see https://github.com/blueyed/pytest/pull/145 (which appears to not be wanted / is expected behavior for pytest itself though, and not really relevant here - just mentioning it since it shows what code is involved there)

A simple implementation to show support for it would probably overwrite the `pytest_collect_file` to not have the check then.
